NebulaNetwork/PacketProcessors/Session/HandshakeRequestProcessor.cs

```diff
@@ -36,27 +39,59 @@ public override void ProcessPacket(HandshakeRequest packet, NebulaConnection con
 
                 pendingPlayers.Remove(conn);
             }
+            
+            Dictionary<string, string> clientMods = new Dictionary<string, string>();
 
+            using (BinaryUtils.Reader reader = new BinaryUtils.Reader(packet.ModsVersion))
+            {
+                for (int i = 0; i < packet.ModsCount; i++)
+                {
+                    string guid = reader.BinaryReader.ReadString();
+                    string version = reader.BinaryReader.ReadString();
+
+                    if (!BepInEx.Bootstrap.Chainloader.PluginInfos.ContainsKey(guid))
+                    {
+                        conn.Disconnect(DisconnectionReason.ModIsMissingOnServer, guid);
+                    }
 
-            if (packet.ModVersion != Config.ModVersion)
+                    clientMods.Add(guid, version);
+                }
+            }
+
+            foreach (var pluginInfo in BepInEx.Bootstrap.Chainloader.PluginInfos)
             {
-                conn.Disconnect(DisconnectionReason.ModVersionMismatch, $"{ packet.ModVersion };{ Config.ModVersion }");
-                return;
+                if (pluginInfo.Value.Instance is IMultiplayerMod mod)
+                {
+                    if (!clientMods.ContainsKey(pluginInfo.Key))
+                    {
+                        conn.Disconnect(DisconnectionReason.ModIsMissing, pluginInfo.Key);
+                        return;
+                    }
+
+                    string version = clientMods[pluginInfo.Key];
+
+                    if (version == mod.Verson || !mod.CheckVersion) continue;
+                    
+                    conn.Disconnect(DisconnectionReason.ModVersionMismatch, $"{pluginInfo.Key};{version};{mod.Verson}");
+                    return;
+                }
             }
 
+
             if (packet.GameVersionSig != GameConfig.gameVersion.sig)
             {
-                conn.Disconnect(DisconnectionReason.GameVersionMismatch, $"{ packet.GameVersionSig };{ GameConfig.gameVersion.sig }");
+                conn.Disconnect(DisconnectionReason.GameVersionMismatch, $"{packet.GameVersionSig};{GameConfig.gameVersion.sig}");
```

Comment:
What is the reason for removing these spaces?

Reply:
This is string interpolation, spaces here mean nothing. It just my IDE formatting the file.